plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/PinotMetadata.java

```diff
@@ -294,6 +314,75 @@ public ConnectorTableProperties getTableProperties(ConnectorSession session, Con
         return Optional.of(new ConstraintApplicationResult<>(handle, remainingFilter, false));
     }
 
+    @Override
+    public Optional<AggregationApplicationResult<ConnectorTableHandle>> applyAggregation(
+            ConnectorSession session,
+            ConnectorTableHandle handle,
+            List<AggregateFunction> aggregates,
+            Map<String, ColumnHandle> assignments,
+            List<List<ColumnHandle>> groupingSets)
+    {
+        if (!isAggregationPushdownEnabled(session)) {
+            return Optional.empty();
+        }
+
+        // Global aggregation is represented by [[]]
+        verify(!groupingSets.isEmpty(), "No grouping sets provided");
+
+        // Pinot currently only supports simple GROUP BY clauses with a single grouping set
+        if (groupingSets.size() != 1) {
+            return Optional.empty();
+        }
+
+        PinotTableHandle tableHandle = (PinotTableHandle) handle;
+        // If aggregate are present than no further aggregations
+        // can be pushed down: there are currently no subqueries in pinot
+        if (tableHandle.getQuery().isPresent() &&
+                !tableHandle.getQuery().get().getAggregateColumns().isEmpty()) {
+            return Optional.empty();
+        }
+
+        ImmutableList.Builder<ConnectorExpression> projections = ImmutableList.builder();
+        ImmutableList.Builder<Assignment> resultAssignments = ImmutableList.builder();
+        ImmutableList.Builder<PinotColumnHandle> aggregationExpressions = ImmutableList.builder();
+
+        for (AggregateFunction aggregate : aggregates) {
+            Optional<PinotColumnHandle> rewriteResult = aggregateFunctionRewriter.rewrite(session, aggregate, assignments, tableHandle);
+            if (rewriteResult.isEmpty()) {
+                return Optional.empty();
+            }
+            PinotColumnHandle pinotColumnHandle = rewriteResult.get();
+            aggregationExpressions.add(pinotColumnHandle);
+            projections.add(new Variable(pinotColumnHandle.getColumnName(), pinotColumnHandle.getDataType()));
+            resultAssignments.add(new Assignment(pinotColumnHandle.getColumnName(), pinotColumnHandle, pinotColumnHandle.getDataType()));
+        }
+        List<String> groupingColumns = getOnlyElement(groupingSets).stream()
```

Comment:
@elonazoulay IIUC the grouping columns in rewrite context are only needed to verify that the grouping columns are part of the SELECT clause. If so, can we instead do that validation here after the rewrite is done?

Yes, the connector will end up doing work that gets discarded but that seems cleaner because even if we add grouping columns to RewriteContext I don't see an easy way to populate them for other connectors so we won't be able to provide a sane impl. which doesn't return `null` for that method.

I think this is possible to do because here we have tableHandle available from which we can get the grouping columns in case of DynamicTable.

Reply:
Was there some specific reason to change from the old implementation? Maybe I'm missing something. If it's feasible to do I'd like that.

As for the RewriteContext changes I'll defer to @martint (and Piotr once he's back). IMO if we have alternatives which are cleaner then I'd like to avoid the change. IMO the validation that grouping columns are part of select isn't exclusive to just count_distinct and makes sense to do before any rewrite.

Reply:
Sounds good. Would another option be to pass the table handle into the rewrite context? I can get the grouping columns there also.

Reply:
@hashhar - I extracted the check for grouping columns and reverted the changes for AggregateFunctionRewriter and RewriteContext. What do you think about adding TableHandle to the RewriteContext instead of this approach?
Just in case I saved the previous version.

Reply:
I like the newer version. We can revisit changing RewriteContext once we have other non-JDBC connectors that implement agg pushdown. We'll have better information at that time to decide the correct abstraction.